### PR TITLE
[DOP-6703] Implement FileReader

### PR DESCRIPTION
**Pull request by dolfinus**
## Change Summary

* Changed `BaseFileDFConnection` class:
  * made `check_if_format_supported` regular method instead of classmethod
  * added `options` to `read_files_as_df`
  * replaced `path: Path` with `paths: list` in `read_files_as_df` - user should be able to read files list
  * added `root: Path` to `read_files_as_df` - for reading list of files and resolving partitioning column values, we have to pass both files list and `basePath` config option to `spark.read`.
* Added `FileReader` class - it has almost the same inputs `FileDownloader`/`FileMover` and part of validation logic, but without handling `source_path` directory content - it just passes all the paths to `connection.read_files_as_df`
* Added `FileReaderOptions` with currently only one option `recursiveFileLookup`. Options like `ignoreMissingFiles` and `ignoreCorruptedFiles` are not supported at the moment - it looks like they are implemented only for ORC/Parquet/Avro, but not CSV. User may pass them though because of `extra = "allow"`, but they are not tested.
* Added fixtures `file_df_connection`. Currently it uses only fixture for `SparkLocalFS`.
* Added tests for `FileReader` with different options. It uses `file_df_connection`, so after fixtures for SparkS3 and SparkHDFS will be added, they will be tested automatically. Only CSV files are tested here.
* Added integration tests for CSV file format with custom options (reading files with header, reading files with custom delimiter, reading files with `inferSchema=true`). More tests will be added here in future PRs.
* Separated documentation of classed related to FileConnection (downloader, uploader, mover, filters, limits) and related to FileDFConnection (reader, formats). See https://onetl--73.org.readthedocs.build/en/73/

## Checklist

* [X] Commit message and PR title is comprehensive
* [X] Keep the change as small as possible
* [X] Unit and integration tests for the changes exist
* [X] Tests pass on CI and coverage does not decrease
* [X] Documentation reflects the changes where applicable
* [ ] `docs/changelog/next_release/<pull request or issue id>.<change type>.rst` file added describing change
  (see [CONTRIBUTING.rst](https://github.com/MobileTeleSystems/onetl/blob/develop/CONTRIBUTING.rst) for details.)
* [X] My PR is ready to review.

### CODE REVIEW

File: onetl/file/file_reader/file_reader.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+                [
+                    # file paths could be relative
+                    "/path/file1.csv",
+                    "/path/nested/file2.csv",
+                ]
+            )
+        """
+
+        if files is None and not self.source_path:
+            raise ValueError("Neither file list nor `source_path` are passed")
```
Comment (dmitry-pedchenko) [inline onetl/file/file_reader/file_reader.py:210]: А мы теперь не пишем то что это FileReader error ?

File: onetl/file/file_reader/file_reader.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+    source_path : os.PathLike or str, optional, default: ``None``
+        Directory path to read data from.
+
+        Could be ``None``, but only if you pass file paths directly to
+        :obj:`~run` method
+
+    df_schema : :obj:`pyspark.sql.types.StructType`, optional, default: ``None``
+        Spark DataFrame schema.
+
+    options : :obj:`onetl.file.file_reader.options.FileReaderOptions`
```
[the remark this answers is not in the document]
Comment (dolfinus) [reply onetl/file/file_reader/file_reader.py]: Добавил

File: onetl/file/file_reader/file_reader.py (hunk 1 above)
Comment (dmitry-pedchenko) [reply onetl/file/file_reader/file_reader.py:210]:
```suggestion
            raise ValueError("|FileReader| Neither file list nor `source_path` are passed")
```
так вот имею виду

**Review by dmitry-pedchenko — APPROVED** [review]

File: onetl/file/file_reader/file_reader.py (hunk 1 above)
Comment (dolfinus) [reply onetl/file/file_reader/file_reader.py:210]: А, из текстов исключений эти префиксы я давно вырезал, они остались только в логах. В исключениях нет смысла их прописывать - их кидает конкретный класс, в stacktrace видно в какой строке вылетело исключение.

**Review by andy-takker — APPROVED** [review]

**Pull request merged**